File: android/media/src/main/java/com/dreampany/media/data/source/memory/MediaProvider.java
package com.dreampany.media.data.source.memory;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteException;
import android.net.Uri;

import com.dreampany.framework.util.CursorUtil;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Maybe;
import timber.log.Timber;

/**
 * Created by Hawladar Roman on 3/8/18.
 * Dreampany Ltd
 * dreampanymail@gmail.com
 */
public abstract class MediaProvider<T> {

    Cursor getCursor() {
        return null;
    }

     T getItem(Cursor cursor) {
        return null;
     }

    List<T> getItems() {
        return getItems(Integer.MAX_VALUE);
    }

    Maybe<List<T>> getItemsRx() {
        return getItemsRx(Integer.MAX_VALUE);
    }

    List<T> getItems(int limit) {
        Cursor cursor = getCursor();
        if (!CursorUtil.hasCursor(cursor)) {
            return null;
        }

        List<T> items = null;
        if (cursor.moveToFirst()) {
            items = new ArrayList<>(cursor.getCount());
            do {
                T item = getItem(cursor);
                if (item != null) {
                    items.add(item);
                }
            } while (items.size() < limit && cursor.moveToNext());
        }
        CursorUtil.closeCursor(cursor);
        return items;
    }

    Maybe<List<T>> getItemsRx(int limit) {
        return Maybe.fromCallable(() -> getItems(limit));
    }

    Cursor resolveQuery(
            Context context,
            Uri uri,
            String[] projection,
            String selection,
            String[] selectionArgs,
            String sortOrder) {

        try {
            return context.getApplicationContext().getContentResolver().query(uri, projection, selection, selectionArgs, sortOrder);
        } catch (SQLiteException invalidQuery) {
            Timber.e("ContentProvider makes exception: %s", invalidQuery.toString());
            return null;
        }
    }
}

File: android/nearby/src/main/java/com/dreampany/nearby/data/source/memory/MediaProvider.kt
package com.dreampany.nearby.data.source.memory

import android.database.Cursor
import com.dreampany.framework.misc.exts.has
import com.dreampany.nearby.data.model.media.Media

/**
 * Created by roman on 28/6/20
 * Copyright (c) 2020 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
abstract class MediaProvider<T : Media> {

    open val cursor: Cursor? = null

    open fun get(cursor: Cursor): T? {
        return null
    }

    open fun gets(): List<T>? {
        val cursor = cursor ?: return null
        if (!cursor.has) {
            return null
        }
        var result: MutableList<T>? = null
        if (cursor.moveToFirst()) {
            result = mutableListOf()
            do {
                val item = get(cursor)
                if (item != null) {
                    result.add(item)
                }
            } while (cursor.moveToNext())
        }
        cursor.close()
        return result
    }
}